crates/fuel-core/src/schema/coins.rs

```diff
@@ -233,75 +298,187 @@ impl CoinQuery {
+    owner: fuel_tx::Address,
+    query_per_asset: Vec<SpendQueryElementInput>,
+    excluded_ids: Option<ExcludeInput>,
+    max_input: u16,
+    base_asset_id: &fuel_tx::AssetId,
+    db: &ReadView,
+) -> async_graphql::Result<Vec<Vec<CoinType>>> {
+    let query_per_asset = query_per_asset
+        .into_iter()
+        .map(|e| {
+            AssetSpendTarget::new(
+                e.asset_id.0,
+                e.amount.0,
+                e.max
+                    .and_then(|max| u16::try_from(max.0).ok())
+                    .unwrap_or(max_input)
+                    .min(max_input),
+            )
+        })
+        .collect_vec();
+    let excluded_ids: Option<Vec<_>> = excluded_ids.map(|exclude| {
+        let utxos = exclude
+            .utxos
+            .into_iter()
+            .map(|utxo| coins::CoinId::Utxo(utxo.into()));
+        let messages = exclude
+            .messages
+            .into_iter()
+            .map(|message| coins::CoinId::Message(message.into()));
+        utxos.chain(messages).collect()
+    });
+
+    let spend_query =
+        SpendQuery::new(owner, &query_per_asset, excluded_ids, *base_asset_id)?;
+
+    let all_coins = random_improve(db, &spend_query)
+        .await?
+        .into_iter()
+        .map(|coins| {
+            coins
+                .into_iter()
+                .map(|coin| match coin {
+                    coins::CoinType::Coin(coin) => CoinType::Coin(coin.into()),
+                    coins::CoinType::MessageCoin(coin) => {
+                        CoinType::MessageCoin(coin.into())
+                    }
+                })
+                .collect_vec()
+        })
+        .collect();
+
+    Ok(all_coins)
 }
 
-impl From<coins::CoinType> for CoinType {
-    fn from(value: coins::CoinType) -> Self {
-        match value {
-            coins::CoinType::Coin(coin) => CoinType::Coin(coin.into()),
-            coins::CoinType::MessageCoin(coin) => CoinType::MessageCoin(coin.into()),
-        }
+async fn into_coin_id(
+    mut selected_stream: impl Stream<Item = CoinsToSpendIndexEntry> + Unpin,
+    max_coins: usize,
+) -> Result<Vec<CoinId>, CoinsQueryError> {
+    let mut coins = Vec::with_capacity(max_coins);
+    while let Some((foreign_key, coin_type)) = selected_stream.next().await {
+        let coin = match coin_type {
+            IndexedCoinType::Coin => {
+                let bytes: [u8; COIN_FOREIGN_KEY_LEN] = foreign_key
+                    .foreign_key_bytes()
+                    .try_into()
+                    .map_err(|_| CoinsQueryError::IncorrectCoinKeyInIndex)?;
+
+                let (tx_id_bytes, output_index_bytes) = bytes.split_at(TxId::LEN);
+                let tx_id = TxId::try_from(tx_id_bytes).map_err(StorageError::from)?;
+                let output_index = u16::from_be_bytes(
+                    output_index_bytes.try_into().map_err(StorageError::from)?,
+                );
+                CoinId::Utxo(fuel_tx::UtxoId::new(tx_id, output_index))
```

Comment:
It would be nice if this conversion logic is a part of the `CoinsToSpendIndexKey`. Either during decoding from the storage you already can decode it and just use the nice typed object(you can check how it was done via `Manual<OwnedTransactionIndexKey>`). Or you can support something like `foreign_key.try_into_utxo_id()?`.

Reply:
The solution with `Manual` is nice, but tricky due to the variable length of the `CoinsToSpendIndexKey`. I would need to specify the `Encoder` associated type as `Vec<u8>`. However, I like the idea of `CoinsToSpendIndexKey` being a struct with specific types, instead of just bytes, but that will incur quite a big rewrite, because I didn't think about it from day 1 :)

I can create a follow-up good-first issue to improve this later.

For now I took the approach of `try_into...()` in this commit: bdc8ce233df9de9a611a00efc93bead03d7fcd20

Reply:
Follow-up issue created: https://github.com/FuelLabs/fuel-core/issues/2498